paddle/framework/executor.cc

```diff
@@ -0,0 +1,164 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#include "paddle/framework/executor.h"
+#include <algorithm>
+#include <iostream>
+#include <memory>
+#include <set>
+#include <vector>
+#include "paddle/framework/lod_tensor.h"
+#include "paddle/framework/op_registry.h"
+#include "paddle/framework/scope.h"
+
+#include <boost/range/adaptor/reversed.hpp>
+
+namespace paddle {
+namespace framework {
+
+Executor::Executor(const std::vector<platform::Place>& places) {
+  device_contexts_.resize(places.size());
+  for (size_t i = 0; i < places.size(); i++) {
+    if (platform::is_cpu_place(places[i])) {
+      device_contexts_[i] = new platform::CPUDeviceContext(
+          boost::get<platform::CPUPlace>(places[i]));
+    } else if (platform::is_gpu_place(places[i])) {
+#ifdef PADDLE_WITH_CUDA
+      device_contexts_[i] = new platform::CUDADeviceContext(
+          boost::get<platform::GPUPlace>(places[i]));
+#else
+      PADDLE_THROW("'GPUPlace' is not supported in CPU only device.");
+#endif
+    }
+  }
+}
+
+Executor::~Executor() {
+  for (auto& device_context : device_contexts_) {
+    if (device_context) {
+      delete device_context;
+    }
+  }
+}
+
+void Executor::Run(const ProgramDesc& pdesc, Scope* scope) {
+  // TODO(tonyyang-svail):
+  //    - only runs the first block
+  //    - only runs on the first device
+  //    - test on gpu
+  auto& block = pdesc.blocks(0);
+  auto& device = device_contexts_[0];
+
+  // TODO(tonyyang-svail):
+  //    - runs on a new local scope
+  // Scope& local_scope = scope->NewScope();
+
+  for (auto& var : block.vars()) {
+    scope->NewVar(var.name());
+  }
+
+  std::vector<bool> should_run = Preprocess(pdesc);
+  PADDLE_ENFORCE(should_run.size() == block.ops_size(),
+                 "should_run.size() != block.ops_size()");
+  for (int i = 0; i < should_run.size(); ++i) {
+    if (should_run[i]) {
+      auto op = paddle::framework::OpRegistry::CreateOp(block.ops(i));
+      std::cout << op->DebugString() << std::endl;
+      op->Run(*scope, *device);
+    }
+  }
+
+  // // print tensor value
+  // for (auto& var : block.vars()) {
+  //   std::cout << var.name() << std::endl;
+  //   auto v = scope->FindVar(var.name());
+  //   const LoDTensor& t = v->Get<LoDTensor>();
+  //   for (int i = 0; i < t.numel(); ++i) {
+  //     std::cout << t.data<float>()[i] << " ";
+  //   }
+  //   std::cout << std::endl;
+  // }
+}
+
+std::vector<bool> Executor::Preprocess(const ProgramDesc& pdesc) {
```

Comment:
Is this what `prune` does? From my understanding `Executor` just runs whatever passed to it. The "prune" should be done by another module, before calling `Executor.Run`.

Reply:
I agree. This function will be moved in the end.

Reply:
Can we implement another class now? Otherwise it violates [Single responsibility principle](https://en.wikipedia.org/wiki/Single_responsibility_principle). It's just a wrapper for the existing code. Probably only a 20 min work :)
If we don't have time now, we probably never have time in the future ;p

Reply:
Talked with @tonyyang-svail offline, it's more effort than I thought, let's do it after the PR gets merged, since this PR is already getting very big.